feature: dune show targets and dune show aliases

Alizter opened the pull request:
Add new `dune show targets` and `dune show aliases` commands that work similarly to `ls` displaying targets and aliases respectively in the current directory or instead any directories given as arguments. They do not show targets or aliases in subdirectories or inside directory targets. Directory targets will be printed with a trailing path separator to distinguish them however.

- fixes https://github.com/ocaml/dune/issues/265

and continuing
- https://github.com/ocaml/dune/pull/1755

part of general effort in:
- #7784 

This is quite useful in cram tests to make sure dune is willing to produce the correct targets.

- [x] changelog
- [x] docs (more cli docs too)
- [x] display directory targets ending with an `/`

Alizter commented:
@emillon Since this is a relatively straightforward addition, it could perhaps be considered for 3.8.

Alizter commented:
@nojb I have something similar giving hints to aliases: https://github.com/ocaml/dune/pull/7004. However due to the way we lookup aliases we have to do it for all contexts at the same time.

Alizter commented:
@nojb Actually I went ahead and implemented an `--aliases` option that shows aliases in addition. Is this what you had in mind?

I'm actually unsure if we want this in the targets command or we wish to have a separate command for displaying aliases. In any case if we are undecided we can always drop that commit.

nojb commented:
> @nojb Actually I went ahead and implemented an `--aliases` option that shows aliases in addition. Is this what you had in mind?

I think so, yes, thanks! What I had in mind is a beginner; I would imagine most of the time such an user will be more interested in invoking an alias than a target file.

snowleopard commented:
That's a very useful command but I feel like it shouldn't be top-level. Don't we usually put these things under `dune describe`? 

I don't like the name "describe" and would prefer something like `dune show` instead, but that's a detail. I think we can end up polluting the top-level namespace if we keep adding commands like `dune targets`.

I

nojb commented:
> I don't like the name "describe" and would prefer something like `dune show` instead, but that's a detail. I think we can end up polluting the top-level namespace if we keep adding commands like `dune targets`.

`dune show` seems like a good top-level name for the kind of functionality. Why not add `dune show` and put the command under it? (`dune show targets`). Existing commands could be migrated under it as well later: `dune show rules` instead of `dune rules`, `dune show env` instead of `dune printenv`, `dune show workspace` instead of `dune describe workspace`, etc

snowleopard commented:
> `dune show` seems like a good top-level name for this kind of functionality. Why not add `dune show` and put the command under it? (`dune show targets`). Existing commands could be migrated under it as well later: `dune show rules` instead of `dune rules`, `dune show env` instead of `dune printenv`, `dune show workspace` instead of `dune describe workspace`, etc

@nojb Sounds like a great plan to me!

snowleopard commented:
> I'm actually unsure if we want this in the targets command or we wish to have a separate command for displaying aliases. In any case if we are undecided we can always drop that commit.

@Alizter If we go with the `dune show` plan, then I'd prefer aliases to be handled by `dune show aliases`.

Alizter commented:
@rgrinberg We can deprecate the bad name (error) and point to the new good name.

rgrinberg commented:
We shouldn't. The disturbance caused by this deprecation far outweighs whatever benefit can gained from it. Going through a deprecating cycle is a ton of churn for both the developers and the users. At best, we can introduce an alias like we did for `runtest`.

nojb commented:
> We shouldn't. The disturbance caused by this deprecation far outweighs whatever benefit can be gained from it.

Even if we don't aggressively deprecate the existing commands and just alias them under new names, I think that, in terms of CLI usability, it would still be an improvement migrating to `dune show` as a top-level name for all these commands. We could do a "soft" deprecation by not documenting the commands under the old names anymore, so that new users would be directed to the new names only. One could even imagine going one step further (if it is possible, I haven't checked) and remove the old names from the auto-generated `dune --help` page. Even if that is not possible we could clearly indicate in the help page that the `dune show` commands are to be preferred.

Alizter commented:
Regarding this PR: I've added a show subcommand and added aliases and a subcommand of that. So we have `dune show aliases` and `dune show targets`. Both have tests.

I'm drafting this PR for the time being so we can continue to discuss how best to group the commands. My thinking is that we really want to add these to `dune describe` and we want to rename `dune describe` to `dune show`. But let's continue that discussion in the issue.

snowleopard commented:
I just want to say that this feature was requested internally (there was particular interest for `dune show aliases`). I wonder if we can make progress here, somehow. Maybe add these commands both into `dune describe` and `dune show`? Alternatively, add them only to `dune describe` for now and create a ticket to think about a possible migration story towards `dune show`.

snowleopard commented on test/blackbox-tests/test-cases/show/dune-aliases.t/run.t:
```diff
@@ -0,0 +1,113 @@
+  > EOF
+
+  $ dune show aliases
+  .:
```
I think I'd prefer the `dir:` not shown in the output because it makes it harder to write automation around `dune show aliases`. I'd imagine that in most cases users will want to filter it out.

snowleopard replied:
(Also, I think it's just confusing to see the directory in the output, particularly `.:` is a pretty obscure string.)

Alizter replied:
Sure , I just copied thr behaviour of `ls` herr.

snowleopard replied:
Hmm, not sure which `ls` you mean: mine doesn't do that :) (And thanks!)

Alizter replied:
For me `ls` only shows the headers when requesting multiple directories. I think I will use this behaviour since in the multiple case the headers have a use.

snowleopard commented on test/blackbox-tests/test-cases/show/dune-targets-simple.t/run.t:
```diff
@@ -0,0 +1,73 @@
+The command also works with files in the _build directory.
+
+  $ dune show targets _build/default/
+  _build/default:
```
Same for `dune show targets`: I think we shouldn't print out the directory here.

rgrinberg commented:
> I just want to say that this feature was requested internally (there was particular interest for dune show aliases). I wonder if we can make progress here, somehow

Externally, it's easy to make progress. We just follow existing conventions.

Internally, we no longer import the command line from upstream. Nor do we have an existing `show` or `describe` command internally, so we can name this whatever we like. However, we would need to factor this out into something that can be used by both command lines.

snowleopard commented:
@rgrinberg Agreed. I suppose we can move the implementation into the engine: it may be useful not only for CLI but for a future specialised Dune RPC too.

(Maybe we need a separate library for stable functionality built on top of the build engine but not sure it's worth the extra complexity.)

rgrinberg commented:
> it may be useful not only for CLI but for a future specialised Dune RPC too.

We already have a library where we put shared rpc components that are independent of rules. So there's no problem there.

I was going to propose a `dune_cli` library that could include shared components that we can reuse in our cli's.

I personally wouldn't want to put this into the engine.

snowleopard commented:
Both options seem good. I wanted to suggest `dune_cli` too but this particular functionality seems useful as an RPC as well, so maybe we're looking for something more general than just `dune_cli`.

rgrinberg commented:
> I wanted to suggest `dune_cli` too but this particular functionality seems useful as an RPC as well, so maybe we're looking for something more general than just `dune_cli`.

In this concrete case, I would put in RPC. Because the shared components of the cli would already depend on the shared components of the RPC (`dune rpc build`, `dune shutdown`, `dune diagnostics` are all shared and need RPC).

snowleopard commented:
Sounds good to me!

rgrinberg reviewed (changes requested):
> test generated targets (not worth investigating at this time)

What are generated targets?

Alizter commented:
> > test generated targets (not worth investigating at this time)
> 
> What are generated targets?

Nevermind, I was confused about something when I wrote it.

Alizter commented on bin/describe/aliases_cmd.ml:
```diff
@@ -0,0 +1,68 @@
+open Import
+open Stdune
```
apparently not

Alizter replied:
I opened #7994 to fix it tho

rgrinberg reviewed (changes requested):
I started doing some cleanups and realized there's a little too much copy pasting between `aliases` and `targets` for my tastes. Can you factor the code out?

Alizter commented:
@rgrinberg sure, I'll have a look over the weekend.

Alizter commented:
@rgrinberg I've factored the common code out, I'm not so sure about the API yet. I think there is some more cleanup we can do.

rgrinberg commented on bin/describe/ls_like_cmd.mli:
```diff
@@ -0,0 +1,8 @@
+     (   Dune_rules.Main.build_system
+      -> Path.Source.t
+      -> Path.t
+      -> string list Action_builder.t)
```
These arguments need some labeling and/or documentation.

1. What's the Path.Source.t?
2. What's the Path.t?
3. What's the list of strings being returned?

Alizter replied:
They are all hidden details now.

Alizter commented:
I've just added a user error for when a directory is outside of the project.;

snowleopard commented:
Perhaps, `dune show sources` could be useful too: for symmetry, and for making it possible to check which files Dune actually considers sources. (Not always obvious.)

rgrinberg commented:
> Perhaps, dune show sources could be useful too

It could also be distinguished via some visual cue in `dune show targets`. We can still add `dune show sources` of course.

Let's save that for a subsequent PR though.

snowleopard commented:
Sure, happy for this to go to a separate PR.
